### PR TITLE
add caching for static frameworks with resources

### LINKED ISSUE #1975: [META] Cache issues & improvements
This is an issue to keep track of all the cache-related issues.
- [x] When the `autosynthesizing` interfaces for resources feature is enabled, the `tuist cache print-hashes` command fails to run because it tries to hash non-existing files (those are generated when the side effects are executed).
- [ ] The current implementation doesn't support the caching of transitive bundles. We should extend the logic to have support for caching them, and doing the right mutations in the graph when users run the `tuist focus` command.
- [x] `tuist cache warm` fails with file not found when a targetAction.outputPaths or targetAction.outputFileListPaths use environment variables [issue](https://github.com/tuist/tuist/issues/1870) [PR](https://github.com/tuist/tuist/pull/2018)
- [X]  App extensions are failing to build for "linking a dylib which is not safe for use in application extension" which was only a warning without using cache - @andreacipriani  tried on a sample project and couldn't reproduce. You get the expected warning but `tuist cache warm` works as expected
- [x] @RomainBoulay reported that using `latest` as the simulator when warming up the cache might not work if developers have custom simulators in their environment. [PR](https://github.com/tuist/tuist/pull/2016)
- [ ] @RomainBoulay reported that using `tuist focus` may lead to an incorrect project generation when some of the dependent targets are present in the local cache.
- [ ] Address un-needed(?)(TBC) warning: ```Target "XXX" has been linked against ["A", "B"], it is a static product so may introduce unwanted side effects.```
- [x] Missing required module 'XXX' when running unit tests with cached dependencies [issue](https://github.com/tuist/tuist/issues/2023)
- [ ] When developers change the default derived data to be relative to the project's directory, the cache warming fails to resolve the path of the output directory.

## Ideas
- [ ] Debugging why branches of the dependency graph are not cacheable is tricky. I think we can build a `tuist cache graph` that highlights what targets are cacheable and which ones are not.

### PR DESCRIPTION
**Pull request by mstfy**
resolves #1975 (caching of transitive bundles)

Short description
Static frameworks that depend on bundles cannot be cached

Solution 
Make bundles cachable.

Implementation
- Create `CacheBundleBuilder` to build bundles
- Add a new case to `ValueGraphDependency` to represent bundles
- Add `BundleLoader` type to load bundles from cache
- Modify `BuildPhaseGenerator` to add bundles to copy bundle resources build phase
- Reorder project mappers to auto create schemes for bundle targets
- Modify `CacheGraphMutator` to generate project for bundles even if the bundle cached but not the framework

### CODE REVIEW

**Review by aniltaskiran — APPROVED** [review]

File: Sources/TuistCore/NodeLoaders/BundleLoader.swift
Comment (pepibumur) [inline Sources/TuistCore/NodeLoaders/BundleLoader.swift:36]: I'd add a unit test that ensures that this method raises when the bundle doesn't exist.

File: Sources/TuistKit/Cache/CacheController.swift
Comment (pepibumur) [inline Sources/TuistKit/Cache/CacheController.swift:228]: What about adjusting the `build` method in `bundleArtifactBuilder` to return the path to the bundle inside the output directory? Otherwise you have the globbing logic in two places, here, and inside the bundle instance.

File: Sources/TuistKit/GraphMappers/ProjectMapperProvider.swift
Comment (pepibumur) [inline Sources/TuistKit/GraphMappers/ProjectMapperProvider.swift:52]: This is worth adding a unit test [here](https://github.com/tuist/tuist/blob/main/Tests/TuistKitTests/GraphMappers/ProjectMapperProviderTests.swift) to ensure that the mapper is in the right position.

Comment (pepibumur) [reply Sources/TuistKit/GraphMappers/ProjectMapperProvider.swift:52]: Also, what about adding a comment explaining why it has to be in this position?

File: Sources/TuistCache/Cache/CacheRemoteStorage.swift
Comment (pepibumur) [inline Sources/TuistCache/Cache/CacheRemoteStorage.swift:154]: Nitpick: I'd rename this method then to something like `artifactPath`. A bundle is not a framework :)

File: Sources/TuistCore/Graph/GraphTraverser.swift
Comment (fortmarek) [inline Sources/TuistCore/Graph/GraphTraverser.swift:482]: I'd just use anonymous parameter `$0.isPrecompiled` instead of `dep` which does not bring any real value here IMHO

File: Sources/TuistCore/NodeLoaders/BundleLoader.swift
Comment (fortmarek) [inline Sources/TuistCore/NodeLoaders/BundleLoader.swift:32]: does this component need to be `TuistCore`? AFAICT, it can be in `TuistCache` only

File: Sources/TuistGenerator/Generator/BuildPhaseGenerator.swift
Comment (fortmarek) [inline Sources/TuistGenerator/Generator/BuildPhaseGenerator.swift:445]:
```suggestion
                .compactMap { dependency -> PBXFileReference? in
```

File: Sources/TuistKit/Cache/CacheController.swift
Comment (fortmarek) [inline Sources/TuistKit/Cache/CacheController.swift:207]: you can use `try FileHandler.shared.inTemporaryDirectory` instead

File: Sources/TuistKit/Cache/CacheController.swift
Comment (fortmarek) [inline Sources/TuistKit/Cache/CacheController.swift:214]: if you didn't have two different methods as I suggested above, this could also be done in a more compact manner by creating a variable and just passing it to the `build` function after the if-else

File: Tests/TuistKitTests/Cache/CacheControllerTests.swift
Comment (fortmarek) [inline Tests/TuistKitTests/Cache/CacheControllerTests.swift:33]: you'll need to add it to `tearDown` as well.

**Review by fortmarek — COMMENTED** [review]
This seems a lot of work 🤯 Thanks!

Additional question:
could you update one of the fixtures that are used [here](https://github.com/tuist/tuist/blob/main/projects/tuist/features/cache-frameworks.feature), so that it contains a static framework with resources? It'd be great if I could try this out and if it was tested on CI

File: Sources/TuistKit/Cache/CacheController.swift
Comment (mstfy) [reply Sources/TuistKit/Cache/CacheController.swift:228]: Makes sense but `build` method is declared on `CacheArtifactBuilding` protocol. Changing its signature will also affect `CacheFrameworkBuilder` and `CacheXCFrameworkBuilder`

File: Sources/TuistCache/Utilities/CacheBundleBuilder.swift
Comment (mstfy) [reply Sources/TuistCache/Utilities/CacheBundleBuilder.swift:10]: Only the `bundleNotFound` case is used. Deleted other cases

File: Sources/TuistCache/Utilities/CacheBundleBuilder.swift
Comment (mstfy) [reply Sources/TuistCache/Utilities/CacheBundleBuilder.swift:46]: It is injected in `CacheFrameworkBuilder` too

File: Sources/TuistCache/Utilities/CacheBundleBuilder.swift
Comment (mstfy) [reply Sources/TuistCache/Utilities/CacheBundleBuilder.swift:72]: Again this change requires `CacheArtifactBuilding` protocol signature which has two `build` methods to change

File: Sources/TuistCore/Graph/GraphTraverser.swift
Comment (mstfy) [reply Sources/TuistCore/Graph/GraphTraverser.swift:482]: I changed it to keypath version 🚀

File: Sources/TuistCore/NodeLoaders/BundleLoader.swift
Comment (mstfy) [reply Sources/TuistCore/NodeLoaders/BundleLoader.swift:32]: Maybe at some point we might add standalone bundle's as dependencies to targets. Like google maps framework makes you manuelly add its resources bundle to the project?

File: Sources/TuistKit/Cache/CacheController.swift
Comment (mstfy) [reply Sources/TuistKit/Cache/CacheController.swift:214]: As I said it requires to change protocol signature.

File: Sources/TuistCore/NodeLoaders/BundleLoader.swift
Comment (fortmarek) [reply Sources/TuistCore/NodeLoaders/BundleLoader.swift:32]: that's fair but for the time being, I'd leave it in a framework where it's exclusively used at the moment. We can move it in the future if we deem it necessary 🙂

File: Sources/TuistCache/Utilities/CacheBundleBuilder.swift
Comment (fortmarek) [reply Sources/TuistCache/Utilities/CacheBundleBuilder.swift:46]: I think it's fine to make the refactor there, too

File: Sources/TuistKit/Cache/CacheController.swift
Comment (fortmarek) [reply Sources/TuistKit/Cache/CacheController.swift:214]: might be worth doing a refactor of the protocol then - but I'll leave it up to you, it's not necessary 🙂

**Comment by fortmarek** [comment]
btw, to fix the format code check you can run `./fourier format swift --fix`

File: Package.resolved
Comment (mstfy) [reply Package.resolved:36]: I will revert these changes.

File: Package.resolved
Comment (mstfy) [reply Package.resolved:203]: Again this is I believe comes from updating pacakge dependencies. I will revert this too

**Comment by fortmarek** [comment]
@pepibumur could you give this another round of review? I'll do it but can make it only by tomorrow/day after tomorrow. Sorry @mstfy for the hold up 🙂

**Comment by fortmarek** [comment]
@mstfy you'll have to rebase to fix the CI

File: projects/tuist/features/cache-frameworks.feature
Comment (fortmarek) [inline projects/tuist/features/cache-frameworks.feature:56]: It should be Given When Then, so this should be prefixed with `And`

**Review by fortmarek — APPROVED** [review]
I have tried this locally and have not run into any issues, awesome work @mstfy 👏

**Comment by pepibumur** [comment]
Waiting for the CI checks to pass. Once they do, we can merge this one

**Comment by pepibumur** [comment]
@all-contributors add @mstfy for feature

**Comment by pepibumur** [comment]
@all-contributors add @mstfy for code

**Comment by pepibumur** [comment]
@mstfy do you mind pushing another commit to trigger all the CI workflows? I want to make sure it's green before we merge this.

**Comment by fortmarek** [comment]
@mstfy you might need to rebase from `main`. Hopefully, this will get this to a mergeable state 🙏

**Comment by fortmarek** [comment]
Great job @mstfy! 👏

**Pull request merged**
